### PR TITLE
Student Progress Bar

### PR DESCRIPTION
**Pull request by FishyFeeshy**
# Descripción

Adds a feature on the students' progress page that creates a student progress bar. The progress bar consists of several boxes for each problem. The color of the box is based on the score the student received, and the score can be viewed by hovering over the box. 

image

# Comentarios

CC: @carlosabcs 

# Checklist:

- [ ] El código sigue la [guía de
      estilo](https://github.com/omegaup/omegaup/wiki/Coding-guidelines) de
      omegaUp.
- [ ] Se corrieron todas las pruebas y pasaron.
- [ ] Si se está agregando funcionalidad nueva, se agregaron pruebas.
- [ ] Si el cambio es grande (> 200 líneas), hay que intentar partirlo en
      varios pull requests. De preferencia uno para los controladores + phpunit
      y luego otro para la interfaz.

### CODE REVIEW

File: frontend/www/js/omegaup/components/course/StudentProgress.test.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+describe('StudentProgress.vue', () => {
+  it('Should handle scores', async () => {
+    const expectedDate = new Date('1/1/2020, 12:00:00 AM');
+    const wrapper = shallowMount(course_StudentProgress, {
+      propsData: {
+        course: {
+          alias: 'hello',
+        },
+        assignment: [
```
Comment (lhchavez) [inline frontend/www/js/omegaup/components/course/StudentProgress.test.ts:17]: would it be possible to put the type annotation to all entries directly under `propsData`?

```suggestion
        assignment: <omegaup.Assignment[]>[
```

the types should match the ones declared in `StudentProgress.vue`. same for `course` and `student`.

File: frontend/www/js/omegaup/components/course/StudentProgress.vue [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+<style>
+.box {
+  width: 20px;
+  height: 20px;
+  border: 1px solid black;
+  margin: -0.5px;
+}
+
+.bg-green {
+  background: rgb(53, 184, 53);
```
Comment (lhchavez) [inline frontend/www/js/omegaup/components/course/StudentProgress.vue:38]: Can these colors be declared in [`frontend/www/sass/base/_colors.scss`](https://github.com/omegaup/omegaup/blob/master/frontend/www/sass/base/_colors.scss)? 
https://github.com/omegaup/omegaup/pull/4424/files#diff-9565d87813be34c611d8564cf7e6e539R67 has an example of how to do that.

File: frontend/www/js/omegaup/components/course/StudentProgress.vue [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+@Component
+export default class StudentProgress extends Vue {
+  @Prop() course!: types.CourseDetails;
+  @Prop() student!: types.StudentProgress;
+  @Prop() assignments!: omegaup.Assignment[];
+
+  T = T;
+
+  problemScore(problem: number): string {
+    return Math.round(problem) + '%';
```
Comment (lhchavez) [inline frontend/www/js/omegaup/components/course/StudentProgress.vue:69]: nit: the use of [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) is preferred over concatenation with `+`:

```suggestion
    return `{Math.round(problem)}%`;
```

(although this was suggested to be inlined above, so this method can be removed).

File: frontend/www/js/omegaup/components/course/AssignmentDetails.vue [hunk 4]
```diff
@@ -222,7 +222,6 @@ import T from '../../lang';
 import course_ProblemList from './ProblemList.vue';
 import course_ScheduledProblemList from './ScheduledProblemList.vue';
 import DateTimePicker from '../DateTimePicker.vue';
-
```
Comment (carlosabcs) [inline frontend/www/js/omegaup/components/course/AssignmentDetails.vue:225]: Try to restore this, and all the lines in all the files, that have been removed. Probably was your editor that automatically did that deletion.

File: frontend/www/js/omegaup/components/course/StudentProgress.vue [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      <div class="d-flex justify-content-center">
+        <div v-if="student.progress.hasOwnProperty(assignment.alias) == false">
+          {{ T.wordsProblemsUnsolved }}
+        </div>
+        <div
+          v-for="problem in student.progress[assignment.alias]"
+          v-bind:class="getProblemColor(Math.round(problem))"
+          data-toggle="tooltip"
+          data-placement="bottom"
+          v-bind:title="problemScore(problem)"
```
Comment (lhchavez) [inline frontend/www/js/omegaup/components/course/StudentProgress.vue:22]:
```suggestion
          v-bind:title="`{Math.round(problemScore)}%`"
```

this expression is simple enough that it's probably better to inline it.

File: frontend/www/js/omegaup/components/course/StudentProgress.vue [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+      class="score flex-column justify-content-center align-items-center"
+      v-for="assignment in assignments"
+    >
+      <p class="mb-1 text-center">{{ Math.round(score(assignment)) }}%</p>
+      <div class="d-flex justify-content-center">
+        <div v-if="student.progress.hasOwnProperty(assignment.alias) == false">
+          {{ T.wordsProblemsUnsolved }}
+        </div>
+        <div
+          v-for="problem in student.progress[assignment.alias]"
```
Comment (lhchavez) [inline frontend/www/js/omegaup/components/course/StudentProgress.vue:18]: nit:

```suggestion
          v-for="problemScore in student.progress[assignment.alias]"
```

given that this is not a problem, but the problem's score, it might be better to call this `problemScore`.

File: frontend/www/js/omegaup/components/course/StudentProgress.test.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+import expect from 'expect';
+import Vue from 'vue';
+import T from '../../lang';
+import { omegaup } from '../../omegaup';
+
+import course_StudentProgress from './StudentProgress.vue';
+
+describe('StudentProgress.vue', () => {
+  it('Should handle scores', async () => {
+    const expectedDate = new Date('1/1/2020, 12:00:00 AM');
```
Comment (carlosabcs) [inline frontend/www/js/omegaup/components/course/StudentProgress.test.ts:11]: This variable must be deleted as you're not using it.

File: frontend/www/js/omegaup/components/course/StudentProgress.test.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+                problem: 44,
+                problem2: 55,
+              },
+            },
+          ],
+          username: 'student',
+        },
+      },
+    });
+    expect(wrapper.find('div.bg-red'));
```
Comment (carlosabcs) [inline frontend/www/js/omegaup/components/course/StudentProgress.test.ts:44]: Try adding another test to check if the value `44` or `55` appears in the component as you expect.

File: frontend/www/js/omegaup/components/course/StudentProgress.vue [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    </td>
+  </tr>
+</template>
+
+<style>
+.box {
+  width: 20px;
+  height: 20px;
+  border: 1px solid black;
+  margin: -0.5px;
```
Comment (carlosabcs) [inline frontend/www/js/omegaup/components/course/StudentProgress.vue:34]: Is this  `margin` really necessary?

File: frontend/www/js/omegaup/components/course/StudentProgress.vue [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    } else if (problemScore >= 50) {
+      return 'box bg-yellow';
+    } else if (problemScore > 0) {
+      return 'box bg-red';
+    } else {
+      return 'box bg-black';
+    }
+  }
+
+  studentProgressUrl(): string {
```
Comment (carlosabcs) [inline frontend/www/js/omegaup/components/course/StudentProgress.vue:99]: This should be a computed property, instead of a method.
```typescript
get studentProgressUrl(): string {
...
}
```

File: frontend/www/js/omegaup/components/problem/List.vue [hunk 11]
```diff
@@ -260,6 +260,7 @@ import {
   faExclamationTriangle,
   faBan,
 } from '@fortawesome/free-solid-svg-icons';
+import StudentProgress from '../course/StudentProgress.vue';
```
Comment (lhchavez) [inline frontend/www/js/omegaup/components/problem/List.vue:263]:
```suggestion
```

this is not being used here, so it can be safely removed.

File: frontend/www/js/omegaup/components/course/StudentProgress.vue [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+      ) /
+        (Object.values(this.student.progress[assignment.alias]).length * 100)) *
+      100
+    );
+  }
+
+  getProblemColor(problemScore: number): String {
+    if (problemScore > 70) {
+      return 'box bg-green';
+    } else if (problemScore >= 50) {
```
Comment (carlosabcs) [inline frontend/www/js/omegaup/components/course/StudentProgress.vue:90]: What is the best option here to avoid those `else if` @lhchavez ?

File: frontend/www/js/omegaup/course/students.ts [hunk 13]
```diff
@@ -1,14 +1,17 @@
 import course_ViewProgress from '../components/course/ViewProgress.vue';
+import course_StudentProgress from '../components/course/StudentProgress.vue';
 import { OmegaUp } from '../omegaup';
 import { types } from '../api_types';
 import * as api from '../api';
 import * as ui from '../ui';
 import T from '../lang';
 import Vue from 'vue';
+import { data } from 'jquery';
```
Comment (lhchavez) [inline frontend/www/js/omegaup/course/students.ts:9]:
```suggestion
```

would it be possible to avoid using jquery? We (and a large part of the rest of the internet) have been slowly trying to remove it everywhere and wouldn't want to see it be reintroduced.

File: frontend/www/js/omegaup/course/students.ts [hunk 14]
```diff
@@ -1,14 +1,17 @@
 import course_ViewProgress from '../components/course/ViewProgress.vue';
+import course_StudentProgress from '../components/course/StudentProgress.vue';
```
Comment (carlosabcs) [inline frontend/www/js/omegaup/course/students.ts:2]: Here you don't need to import the component, so you could delete this line.

File: frontend/www/js/omegaup/course/students.ts (hunk 13 above)
Comment (carlosabcs) [inline frontend/www/js/omegaup/course/students.ts:9]: Why are you importing this?

File: frontend/www/js/omegaup/course/students.ts [hunk 15]
```diff
@@ -1,14 +1,17 @@
...
 import * as api from '../api';
 import * as ui from '../ui';
 import T from '../lang';
 import Vue from 'vue';
+import { data } from 'jquery';
 
 OmegaUp.on('ready', function () {
   const payload = types.payloadParsers.StudentsProgressPayload();
 
+  //what to do for student progress component?
```
Comment (carlosabcs) [inline frontend/www/js/omegaup/course/students.ts:14]: You don't need to do anything here with that component :)

File: frontend/www/js/omegaup/components/course/AssignmentDetails.vue (hunk 4 above)
Comment (lhchavez) [inline frontend/www/js/omegaup/components/course/AssignmentDetails.vue:225]: would it be possible to avoid these whitespace changes?

These can be safely reverted with

```shell
git restore --version=upstream/master frontend/www/js/omegaup/components/course/AssignmentDetails.vue
```

same with all the other files with only whitespace changes.

File: frontend/www/js/omegaup/components/course/Form.test.ts [hunk 16]
```diff
@@ -12,12 +12,12 @@ const baseCourseFormProps = {
     admission_mode: 'registration',
     alias: 'Newx',
     assignments: [],
-    needs_basic_information: false,
     description: 'New',
     finish_time: new Date(),
     is_admin: true,
     is_curator: true,
     name: 'Nuevo',
+    needs_basic_information: false,
```
Comment (lhchavez) [inline frontend/www/js/omegaup/components/course/Form.test.ts:20]: is this change really needed? if not, can this be reverted too?

File: frontend/www/js/omegaup/components/course/StudentProgress.test.ts (hunk 8 above)
Comment (lhchavez) [inline frontend/www/js/omegaup/components/course/StudentProgress.test.ts:44]: what is the expected list of boxes? would it be possible to make a stronger assertion, like checking that all of the boxes exist, are in the right order, and have the expected percentage?

File: frontend/www/js/omegaup/course/students.ts (hunk 15 above)
Comment (lhchavez) [inline frontend/www/js/omegaup/course/students.ts:14]: @tvanessa @heduenas can answer this probably.

File: frontend/www/js/omegaup/components/course/StudentProgress.vue (hunk 12 above)
Comment (lhchavez) [reply frontend/www/js/omegaup/components/course/StudentProgress.vue:90]: something like this:

```typescript
if (problemScore >= 70) return 'box bg-green';
if (problemScore >= 50) return 'box bg-yellow';
...
return 'box bg-black';
```

File: frontend/www/js/omegaup/components/course/StudentProgress.vue (hunk 9 above)
Comment (FishyFeeshy) [reply frontend/www/js/omegaup/components/course/StudentProgress.vue:34]: If I don't include it, the borders are next to each other instead of shared. Is that ok?
image

Comment (lhchavez) [reply frontend/www/js/omegaup/components/course/StudentProgress.vue:34]: if the borders are made a bit less dark, AND the container also has a 1px border, it would make all the boxes have a consistent 2px border. 

This is how the demo in the Bootstrap docs for flex does it, and looks alright: https://getbootstrap.com/docs/4.5/utilities/flex/#direction

**Comment by lhchavez** [comment]
oh forgot to ask one more thing! is there a way in which this can be displayed in a way that is friendly for people with [color vision deficiency](https://en.wikipedia.org/wiki/Color_blindness)?

maybe putting the number in the box with enough contrast suffices. (@carlosabcs @tvanessa for second opinion)

File: frontend/www/js/omegaup/components/course/StudentProgress.vue (hunk 9 above)
Comment (FishyFeeshy) [reply frontend/www/js/omegaup/components/course/StudentProgress.vue:34]: This should be better. 
image

File: frontend/www/js/omegaup/components/course/StudentProgress.vue (hunk 10 above)
Comment (FishyFeeshy) [reply frontend/www/js/omegaup/components/course/StudentProgress.vue:99]: Should this be changed to `return /course/${this.course.alias}/student/${this.student.name || this.student.username}/;`, in case the student doesn't have a name?

Comment (lhchavez) [reply frontend/www/js/omegaup/components/course/StudentProgress.vue:99]: oh! we all missed this. it should be

```typescript
    return `/course/${this.course.alias}/student/${this.student.username}/`;
```

URLs should never use names: instead they should use usernames or aliases (or IDs in very rare cases, like for Schools).

File: frontend/www/js/omegaup/components/course/StudentProgress.vue [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+        <div v-if="student.progress.hasOwnProperty(assignment.alias) == false">
+          {{ T.wordsProblemsUnsolved }}
+        </div>
+        <div v-else class="d-flex border border-dark">
+          <div
+            v-for="problemScore in student.progress[assignment.alias]"
+            v-bind:class="getProblemColor(Math.round(problemScore))"
+            data-toggle="tooltip"
+            data-placement="bottom"
+            v-bind:title="Math.round(problemScore) + '%'"
```
Comment (lhchavez) [inline frontend/www/js/omegaup/components/course/StudentProgress.vue:23]: nit: the use of [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) is preferred over concatenation with `+`:

```suggestion
            v-bind:title="`{Math.round(problemScore)}%`"
```

Comment (FishyFeeshy) [reply frontend/www/js/omegaup/components/course/StudentProgress.vue:23]: When I do that, I get this:
image
Am I doing that wrong? I thought you couldn't do [template literals inside attributes](https://vuejs.org/v2/guide/syntax.html#Attributes)

Comment (carlosabcs) [reply frontend/www/js/omegaup/components/course/StudentProgress.vue:23]: Change that to:
```js
v-bind:title="`${Math.round(problemScore)}%`"
```
and it'll work

Comment (lhchavez) [reply frontend/www/js/omegaup/components/course/StudentProgress.vue]: errr. yes, sorry, i missed a `$` ^^;;

**Comment by lhchavez** [comment]
@carlosabcs any final comments?

@FishyFeeshy there's one conflicting file. you might need to do `git pull upstream master` and resolve the conflicts before reuploading.

File: frontend/www/js/omegaup/components/course/StudentProgress.vue [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+        {{ student.name || student.username }}
+      </a>
+    </td>
+    <td
+      class="score flex-column justify-content-center align-items-center"
+      v-for="assignment in assignments"
+    >
+      <p class="mb-1 text-center">{{ Math.round(score(assignment)) }}%</p>
+      <div class="d-flex justify-content-center">
+        <div v-if="student.progress.hasOwnProperty(assignment.alias) == false">
```
Comment (carlosabcs) [inline frontend/www/js/omegaup/components/course/StudentProgress.vue:14]: This could be modified to: `v-if="!student.progress.hasOwnProperty(assignment.alias)"`
